### PR TITLE
Add sca package scan

### PR DESCRIPTION
**Pull request by gruebel**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

- added new dependency to make pretty tables 😄 
- added new attribute `vulnerabilities`  to `Record` to be able to print it out better
- divided SCA runner in 3 parts -> `runner`, `scanner` and `output`

### CODE REVIEW

File: checkov/sca_package/output.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        file_abs_path=file_abs_path,
+        severity=severity,
+        description=description,
+        short_description=f"{cve_id} - {package_name}: {package_version}",
+        vulnerability_details=details,
+    )
+    return record
+
+
+def calculate_lowest_complaint_version(
```
Comment (nimrodkor) [inline checkov/sca_package/output.py:84]: Complaint - something is wrong
Compliant - upholds a certain rule / benchmark :)
```suggestion
def calculate_lowest_compliant_version(
```

File: checkov/sca_package/output.py [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    fix_versions_lists: List[List[Union[packaging_version.Version, packaging_version.LegacyVersion]]]
+) -> str:
+    """A best effort approach to find the lowest complaint version"""
```
Comment (nimrodkor) [inline checkov/sca_package/output.py:87]:
```suggestion
    """A best effort approach to find the lowest compliant version"""
```

File: checkov/sca_package/output.py [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+                        "id": record.vulnerability_details["id"],
+                        "severity": record.severity,
+                        "fixed_version": record.vulnerability_details["lowest_fixed_version"],
+                    }
+                )
+
+            if package_name in package_details_map.keys():
+                package_details_map[package_name]["cves"].sort(key=compare_cve_severity)
+                package_details_map[package_name]["current_version"] = package_version
+                package_details_map[package_name]["complaint_version"] = calculate_lowest_complaint_version(
```
Comment (nimrodkor) [inline checkov/sca_package/output.py:171]:
```suggestion
                package_details_map[package_name]["compliant_version"] = calculate_lowest_compliant_version(
```

File: checkov/sca_package/output.py [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    package_table_lines: List[str] = []
+    package_table = PrettyTable(min_table_width=table_width, max_table_width=table_width)
+    package_table.set_style(SINGLE_BORDER)
+    package_table.field_names = [
+        "Package",
+        "CVE ID",
+        "Severity",
+        "Current version",
+        "Fixed version",
+        "Complaint version",
```
Comment (nimrodkor) [inline checkov/sca_package/output.py:245]:
```suggestion
        "Compliant version",
```

**Review by nimrodkor — COMMENTED** [review]
Very nice!

File: checkov/sca_package/scanner.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+    async def run_scan_multi(
+        self,
+        address: str,
+        bc_api_key: str,
+        input_output_paths: "Iterable[Tuple[Path, Path]]",
+    ) -> "Sequence[Dict[str, Any]]":
+        args = [
+            (
+                f"./{TWISTCLI_FILE_NAME} coderepo scan --address {address} --token {bc_api_key} --output-file {output_path.absolute()} {input_path.absolute()}",
```
Comment (Saarett) [inline checkov/sca_package/scanner.py:38]: Are we still assuming the users have this file? Should we download it for them if the don't?

**Review by Saarett — COMMENTED** [review]
Absolutely amazing job! 🎩 
Added some comments in the code :)

Regarding the tests, I think we should create more tests for this new scanner as it's going to cover a lot of different package managers and file. Perhaps we should add a test for package-lock.json too, and edge cases like - what if the package management file is empty?

One way or another, I love this work :)

File: checkov/common/output/report.py [hunk 6]
```diff
@@ -20,6 +21,32 @@
 
 init(autoreset=True)
 
+CHECK_TYPE_SCA_PACKAGE = "sca_package"
```
[the remark this answers is not in the document]
Comment (gruebel) [reply checkov/common/output/report.py:24]: good catch, I used it before I created the `dataclass`

File: checkov/main.py [hunk 7]
```diff
@@ -53,12 +53,12 @@
 logging_init()
 logger = logging.getLogger(__name__)
 checkov_runners = ['cloudformation', 'terraform', 'kubernetes', 'serverless', 'arm', 'terraform_plan', 'helm',
-                   'dockerfile', 'secrets', 'json', 'github_configuration', 'gitlab_configuration']
+                   'dockerfile', 'secrets', 'json', 'github_configuration', 'gitlab_configuration', 'sca_package']
```
[the remark this answers is not in the document]
Comment (gruebel) [reply checkov/main.py:56]: changed it to some `dataclass` wizardy 😄

File: checkov/sca_package/output.py (hunk 1 above)
Comment (gruebel) [reply checkov/sca_package/output.py:84]: 🙈

File: checkov/sca_package/output.py (hunk 4 above)
Comment (gruebel) [reply checkov/sca_package/output.py:245]: when you do it wrong, then at least it is consistent 😂

File: checkov/sca_package/output.py [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+            if isinstance(version, packaging_version.Version) and version.major == package_max_version.major
+        )
+        return str(lowest_version)
+
+
+def compare_cve_severity(cve: Dict[str, str]) -> int:
+    return SEVERITY_RANKING[cve.get("severity") or DEFAULT_SEVERITY]
+
+
+def create_cli_output(*cve_records: List[Record]) -> str:
```
[the remark this answers is not in the document]
Comment (gruebel) [reply checkov/sca_package/output.py:117]: as you found it, during the report output creation

File: checkov/sca_package/scanner.py (hunk 5 above)
Comment (nimrodkor) [reply checkov/sca_package/scanner.py:38]: We should wrap the `{input_path.absolute()}` with `"` to handle spaces correctly @gruebel

File: checkov/common/output/report.py [hunk 9]
```diff
@@ -206,20 +233,26 @@ def print_console(
...
-        for record in self.failed_checks:
-            print(record.to_string(compact=is_compact, use_bc_ids=use_bc_ids))
-        if not is_quiet:
-            for record in self.skipped_checks:
+
+        # output for vulnerabilities is different
+        if self.check_type == CheckType.SCA_PACKAGE:
+            if self.failed_checks or self.skipped_checks:
+                from checkov.sca_package.output import create_cli_output  # circular dependency
+                print(create_cli_output(self.failed_checks, self.skipped_checks))
```
[the remark this answers is not in the document]
Comment (gruebel) [reply checkov/common/output/report.py:246]: surprisingly it works, I thought it will blow up 😄 if it results in any errors in the future, then we can just move it back

File: checkov/common/output/report.py [hunk 10]
```diff
@@ -307,6 +346,22 @@ def get_sarif_json(self, tool) -> Dict[str, Any]:
                     }
                 ],
             }
+
+            if record.check_result.get("result") == CheckResult.SKIPPED:
+                # sca_package suppressions can only be enabled via flag
```
[the remark this answers is not in the document]
Comment (gruebel) [reply checkov/common/output/report.py:351]: a SARIF report has a dedicated way on how to inform the user about suppression, it is added to the result under `suppressions` there you can add extra information like `justification`, which is our `suppress_comment` and a `kind`, which can have two different values `inSource` or `external`. Since we currently suppress/skip checks for IaC only via skip comment directly in the code, I default it to `inSource`. Using `--skip-check` for IaC results in not checking it at all, which is different for SCA and therefore it defaults to `external`

File: checkov/sca_package/scanner.py (hunk 5 above)
Comment (gruebel) [reply checkov/sca_package/scanner.py:38]: the file is always downloaded

File: checkov/sca_package/output.py [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+    for file_path, packages in group_by_file_path_package_map.items():
+        cve_count = {
+            "total": 0,
+            "critical": 0,
+            "high": 0,
+            "medium": 0,
+            "low": 0,
+            "skipped": 0,
+            "fixable": 0,
+            "to_fix": 0,
```
[the remark this answers is not in the document]
Comment (gruebel) [reply checkov/sca_package/output.py:135]: created a `dataclass`

**Review by nimrodkor — APPROVED** [review]

**Comment by gruebel** [comment]
@Saarett in general we should add more tests, would need to differentiate between real unit tests and integration tests (interacting with `twistcli`) best would be to create a separate issue and list the test cases, which we want to cover.

**Review by Saarett — APPROVED** [review]
💯
